File: scripts/z-sheet-fill.mjs
// Z曲げ実績記入シートの空欄を、シミュレーター（干渉判定そのもの）で埋めるための計算。
// 入力: [{row, V, t, mat, minOut, actI, actK}] の JSON ファイル
// 出力: 行ごとの sim 値（段差S最小・フランジA上限・A最大・フランジB最小／最大 ほか）
//
// 形の約束（記入シートの図と同じ）:
//   辺 = [フランジA, 段差S, フランジB]（すべて外寸）。1工程目にA側、2工程目は裏返してB側を曲げる。
//
// 使い方: node scripts/z-sheet-fill.mjs <rows.json>
import { readFileSync } from 'node:fs';

const lines = readFileSync('bending-simulator.jsx', 'utf8').split(/\r?\n/);
const end = lines.findIndex((l) => l.startsWith('function FinishedPreview'));
const E = new Function(
  lines.slice(0, end).filter((l) => !l.startsWith('import ')).join('\n') +
  '\nreturn {resolveDie,computeChain,minGap,shoulderReach,reachCheck,toolsFor,strokeState,simMinStep,lookupTable,NOBI_TABLE};'
)();

// V幅 → 実機の型と機械（V.dxf で確認した組み合わせ）
const SEL = {
  8: 'ins:971561:stack', 12: 'ins:974061:stack', 16: 'ins:977061:stack', 20: 'ins:979061:stack',
  25: 'ins:982061:stack', 32: 'lib:03500:0', 40: 'lib:03600:0', 50: 'lib:03700:0', 63: 'lib:03800:0',
  80: 'lib:01360:0', 100: 'lib:01860:0', 125: 'lib:03900:0', 160: 'lib:01400:0',
};
const machineOf = (V) => (V <= 25 ? 'hg2203' : 'hd3504nt');
const PUNCH = '904061';
const HI = 400;            // これで通れば「なし（上限なし）」
const BIG_S = 300;         // フランジA最大を見るときの段差（台に掛からない広さ）

const rows = JSON.parse(readFileSync(process.argv[2], 'utf8'));
const out = [];

for (const r of rows) {
  const sel = SEL[r.V];
  if (!sel) { out.push({ row: r.row, why: '金型なし' }); continue; }
  const machine = machineOf(r.V);
  const info = E.resolveDie(sel, r.V, 30, true, machine);
  const vHalf = info.vHalf;
  const nb = E.lookupTable(E.NOBI_TABLE, r.mat, r.V, r.t);
  if (!nb) { out.push({ row: r.row, why: '片伸びが表に無い' }); continue; }
  const nobi = nb.val, grow = nobi - r.t / 2;
  const openGap = Math.max(0, 170 - r.t);
  const exArc = E.shoulderReach(vHalf, r.t, 90) + r.t;
  const minOut = r.minOut || 0;
  // 相手側のフランジ（固定する側）。折り曲げ表の最小外寸を必ず満たす長さにする。
  // V肩に届く長さ（外寸）より短いと、それだけで曲げられないので、それ以上にする。
  const reachOuter = E.shoulderReach(vHalf, r.t, 90) + nobi + 2;
  const other = Math.max(30, minOut + 5, reachOuter);

  const seq = [{ bend: 0, mirror: false, valley: false }, { bend: 1, mirror: false, valley: true }];
  const feasible = (A, S, B) => {
    if (A < minOut - 1e-6 || B < minOut - 1e-6) return false;      // 折り曲げ表の最小外寸は守る
    const segs = [A - nobi, S - 2 * nobi, B - nobi];
    if (segs.some((x) => x <= 0.5)) return false;
    const part = { t: r.t, segs, bends: [{ angle: 90, dir: 1 }, { angle: 90, dir: -1 }], grow: [grow, grow] };
    for (let si = 0; si < seq.length; si++) {
      if (!E.reachCheck(part, seq, si, vHalf).ok) return false;
      for (let p = 0; p <= 1.0001; p += 0.05) {
        const { bendProg, lift } = E.strokeState(p, openGap);
        const ch = E.computeChain(part, seq, si, bendProg, vHalf);
        if (!ch.activeDirOK) return false;
        const tools = E.toolsFor(PUNCH, false, ch.innerY - lift, 'std', info.polys);
        if (E.minGap(ch, tools.polys, r.t, vHalf, exArc).gap < -0.05) return false;
      }
    }
    return true;
  };
  // 上限を二分探索（lo で通ることが前提）
  const upper = (f, lo) => {
    if (!f(lo)) return { v: null, why: '通らない' };
    if (f(HI)) return { v: 'なし' };
    let a = lo, b = HI;
    for (let i = 0; i < 13; i++) { const m = (a + b) / 2; if (f(m)) a = m; else b = m; }
    return { v: +a.toFixed(1) };
  };
  // 下限を二分探索（hi で通ることが前提）
  const lower = (f, hi) => {
    if (!f(hi)) return { v: null, why: '通らない' };
    let a = 1, b = hi;
    if (f(a)) return { v: +a.toFixed(1) };
    for (let i = 0; i < 13; i++) { const m = (a + b) / 2; if (f(m)) b = m; else a = m; }
    return { v: +b.toFixed(1) };
  };

  // --- 段差S 最小（両端フランジ 80、表の最小外寸が大きい型はその値＋15）
  const flangeOuter = Math.max(80, minOut + 15);
  const part0 = { t: r.t, segs: [flangeOuter - nobi, Math.max(5, (r.actK || 40) - 2 * nobi), flangeOuter - nobi],
                  bends: [{ angle: 90, dir: 1 }, { angle: 90, dir: -1 }], grow: [grow, grow] };
  let simK = null, simKwhy = '';
  try {
    const flat = E.simMinStep(part0, seq, 1, vHalf, info.polys, PUNCH, false, 'std', openGap);
    if (flat == null) simKwhy = '段差を広げても不可'; else simK = +(flat + 2 * nobi).toFixed(1);
  } catch (e) { simKwhy = 'err:' + e.message; }
  // simMinStep は探す範囲が狭く、V100 以上のように段差を大きくとる型では「不可」と返してしまう。
  // そのときは 400mm まで広げて自分で探し直す（400mm で通れば、通る最小を二分探索）。
  if (simK == null) {
    const g = (S) => feasible(flangeOuter, S, flangeOuter);
    if (g(HI)) {
      let a = 2 * nobi + 1, b = HI;
      for (let i = 0; i < 14; i++) { const m = (a + b) / 2; if (g(m)) b = m; else a = m; }
      simK = +b.toFixed(1); simKwhy = '';
    } else {
      simKwhy = '段差を400mmまで広げても不可';
    }
  }

  // 以降の欄で使う段差：実績があれば実績、無ければ sim の最小段差。
  // sim の最小段差ちょうどは境目なので、相手フランジを変えると通らないことがある。
  // そのときは 0.5mm ずつ広げて、実際に通る段差を使う。
  let S = r.actK != null ? r.actK : simK;
  const Ssrc = r.actK != null ? '実績' : 'sim';
  if (Ssrc === 'sim' && S != null) {
    const loA0 = Math.max(minOut, E.shoulderReach(vHalf, r.t, 90) + nobi + 2, other);
    let s2 = S;
    while (s2 <= S + 40 && !feasible(loA0, s2, other)) s2 = +(s2 + 0.5).toFixed(1);
    S = s2 <= S + 40 ? s2 : null;
  }
  const res = { row: r.row, die: info.note.split('｜')[0].trim(), machine, nobi, flangeOuter,
                simK, simKwhy, S, Ssrc };
  if (S == null) { out.push({ ...res, why: '段差が決まらない' }); continue; }

  const loA = Math.max(minOut, E.shoulderReach(vHalf, r.t, 90) + nobi + 2);
  const I = upper((A) => feasible(A, S, other), loA);               // その段差でのフランジA上限
  const J = upper((A) => feasible(A, BIG_S, other), loA);           // 段差を十分とったときのフランジA最大
  const L = lower((B) => feasible(other, S, B), Math.max(other, loA)); // その段差でのフランジB最小
  const M = upper((B) => feasible(other, S, B), Math.max(other, loA)); // その段差でのフランジB最大
  out.push({ ...res, other, simI: I.v, simIwhy: I.why || '', simJ: J.v, simJwhy: J.why || '',
             simL: L.v, simLwhy: L.why || '', simM: M.v, simMwhy: M.why || '' });
}
process.stdout.write(JSON.stringify(out));

File: scripts/z-sheet-fill.py
# -*- coding: utf-8 -*-
# Z曲げ実績記入シートの「実測以外の空欄」をシミュレーション値で埋める。
#   ・実測（I〜N に入っている値）には一切触らない。
#   ・空いている I〜M にシミュ値を入れる。青い斜体＋セルのコメント「シミュレーション値」で区別。
#   ・R〜X（シミュ列）は、いまの土台（V.dxf で確認済み）で計算し直して上書きする。
#   ・元のファイルは上書きせず、別名で保存する。
#
# 使い方: python scripts/z-sheet-fill.py <入力xlsx> <出力xlsx>
import sys, json, subprocess, os, datetime
import openpyxl
from openpyxl.styles import Font, PatternFill
from openpyxl.comments import Comment

src, dst = sys.argv[1], sys.argv[2]
wb = openpyxl.load_workbook(src)
ws = wb.active

rows = []
for r in range(9, 52):
    V, t, mat = ws[f"B{r}"].value, ws[f"E{r}"].value, ws[f"D{r}"].value
    if not isinstance(V, (int, float)) or not isinstance(t, (int, float)):
        continue
    num = lambda c: ws[f"{c}{r}"].value if isinstance(ws[f"{c}{r}"].value, (int, float)) else None
    rows.append({"row": r, "V": int(V), "t": float(t), "mat": mat, "minOut": num("F"),
                 "actI": num("I"), "actK": num("K")})

tmp = os.path.join(os.environ.get("TEMP", "."), "zfill_rows.json")
json.dump(rows, open(tmp, "w", encoding="utf-8"), ensure_ascii=False)
res = subprocess.run(["node", "scripts/z-sheet-fill.mjs", tmp], capture_output=True, text=True, encoding="utf-8")
if res.returncode != 0:
    print(res.stderr); sys.exit(1)
sims = {o["row"]: o for o in json.loads(res.stdout)}

SIM_FONT = Font(color="1F4E9E", italic=True)
SIM_FILL = PatternFill("solid", fgColor="E8F0FC")
today = datetime.date.today().isoformat()

def put_sim(cell, value, note):
    cell.value = value
    cell.font = SIM_FONT
    cell.fill = SIM_FILL
    cell.comment = Comment(f"シミュレーション値（{today}）\n{note}", "sim")

filled = 0
for r, o in sims.items():
    if "why" in o and "simI" not in o:
        reason = o.get("simKwhy") or o["why"]
        if ws[f"K{r}"].value in (None, ""):
            put_sim(ws[f"K{r}"], "算出できず", f"{reason}（両端フランジ {o.get('flangeOuter')} mm）")
            filled += 1
        ws[f"R{r}"] = "算出できず"; ws[f"U{r}"] = reason
        continue
    cond = f"{o['die']}／ヤゲン904061／{'段差S=実績' if o['Ssrc']=='実績' else '段差S=sim最小'} {o['S']}／相手フランジ{o['other']}"
    # --- 記入欄（実測が空のところだけ）
    targets = [
        ("K", o["simK"], o["simKwhy"] or "両端フランジ %s mm で段差を詰めた最小" % o["flangeOuter"]),
        ("I", o["simI"], o["simIwhy"] or "その段差SでのフランジA上限。" + cond),
        ("J", o["simJ"], o["simJwhy"] or "段差を300mmとったときのフランジA上限（なし＝400mmでも当たらない）"),
        ("L", o["simL"], o["simLwhy"] or "その段差SでのフランジB最小。" + cond),
        ("M", o["simM"], o["simMwhy"] or "その段差SでのフランジB上限（なし＝400mmでも当たらない）。" + cond),
    ]
    for col, val, note in targets:
        c = ws[f"{col}{r}"]
        if c.value not in (None, ""):
            continue                       # 実測は触らない
        put_sim(c, val if val is not None else note[:20], note)
        filled += 1
    # --- シミュ列（R〜X）を今の土台で計算し直す
    actK = next((x["actK"] for x in rows if x["row"] == r), None)
    actI = next((x["actI"] for x in rows if x["row"] == r), None)
    ws[f"R{r}"] = o["simK"] if o["simK"] is not None else "算出できず"
    ws[f"S{r}"] = max(v for v in (o["simK"], actK) if v is not None) if (o["simK"] is not None or actK is not None) else None
    ws[f"T{r}"] = round(actK - o["simK"], 1) if (actK is not None and o["simK"] is not None) else None
    ws[f"U{r}"] = cond
    ws[f"V{r}"] = o["simI"]
    if actI is not None and isinstance(o["simI"], (int, float)):
        d = round(actI - o["simI"], 1)
        ws[f"W{r}"] = d
        ws[f"X{r}"] = "ほぼ一致" if abs(d) <= 5 else ("シミュレーションが甘い（危険）" if d < 0 else "シミュレーションが厳しめ")
    else:
        ws[f"W{r}"] = None; ws[f"X{r}"] = None

ws["R6"] = f"シミュレーション（{today} 再計算・V.dxf で確認した土台）"
ws["V6"] = f"フランジA上限の突き合わせ（{today}）"

# 注記
last = ws.max_row + 2
notes = [
    f"※ {today}：実測の入っていない I〜M 列を、シミュレーション値で埋めました（青い斜体・セルにコメントあり）。実測値ではありません。",
    "※ 段差S最小（K）が空の行は、その sim 値を使って I・L・M を計算しています。実測の K がある行は実測の段差で計算しています。",
    "※ フランジA最大（J）は段差を300mmとったときの上限、「なし」は400mmまで当たらないという意味です。",
    "※ R〜X 列は、V.dxf で確認した土台（HG2203 は V8〜V25 共通の台）で計算し直しました。",
]
for i, n in enumerate(notes):
    ws[f"A{last + i}"] = n
    ws[f"A{last + i}"].font = Font(color="1F4E9E")

wb.save(dst)
print("書き出し", dst, "／ 記入したセル", filled)
json.dump(list(sims.values()), open(os.path.join(os.environ.get("TEMP", "."), "zfill_result.json"), "w", encoding="utf-8"),
          ensure_ascii=False, indent=1)
